src/main/java/de/tum/in/www1/artemis/service/FileService.java

```diff
@@ -101,6 +129,84 @@ public void resetOnPath(String path) {
+     * @return The path of the file
+     */
+    @NotNull
+    public String handleSaveFile(MultipartFile file, boolean keepFileName, boolean markdown) {
+        // check for file type
+        String filename = file.getOriginalFilename();
+        if (filename == null) {
+            throw new IllegalArgumentException("Filename cannot be null");
+        }
+        // sanitize the filename and replace all invalid characters with "_"
+        filename = filename.replaceAll("[^a-zA-Z\\d\\.\\-]", "_");
+        String fileExtension = FilenameUtils.getExtension(filename);
+        if (this.allowedFileExtensions.stream().noneMatch(fileExtension::equalsIgnoreCase)) {
+            throw new BadRequestAlertException("Unsupported file type! Allowed file types: " + String.join(", ", this.allowedFileExtensions), "file", null, true);
+        }
+
+        final String filePath = markdown ? FilePathService.getMarkdownFilePath() : FilePathService.getTempFilePath();
+        final String fileNameAddition = markdown ? "Markdown_" : "Temp_";
+        final StringBuilder responsePath = new StringBuilder(markdown ? MARKDOWN_FILE_SUBPATH : DEFAULT_FILE_SUBPATH);
+
+        try {
+            File newFile = createNewFile(filePath, filename, fileNameAddition, fileExtension, keepFileName);
+            responsePath.append(newFile.toPath().getFileName());
+
+            // copy contents of uploaded file into newly created file
+            Files.copy(file.getInputStream(), newFile.toPath(), REPLACE_EXISTING);
+
+            return responsePath.toString();
+        }
+        catch (IOException e) {
+            log.error("Could not save file {}", filename, e);
+            throw new InternalServerErrorException("Could not create file");
+        }
+    }
+
+    /**
+     * Creates a new file from given contents
+     * @param filePath the path to save the file to excluding the filename
+     * @param filename the filename of the file to save
+     * @param fileNameAddition the addition to the filename to make sure it is unique
+     * @param fileExtension the extension of the file to save
+     * @param keepFileName specifies if original file name should be kept
+     * @return the created file
+     */
+    private File createNewFile(String filePath, String filename, String fileNameAddition, String fileExtension, boolean keepFileName) throws IOException {
+        try {
+            Files.createDirectories(Paths.get(filePath));
+        }
+        catch (IOException e) {
+            log.error("Could not create directory: {}", filePath);
+            throw e;
+        }
+        boolean fileCreated;
+        File newFile;
+        String newFilename = filename;
+        do {
+            if (!keepFileName) {
+                newFilename = fileNameAddition + ZonedDateTime.now().toString().substring(0, 23).replaceAll(":|\\.", "-") + "_" + UUID.randomUUID().toString().substring(0, 8) + "."
```

Comment:
```suggestion
                // append a timestamp up to millisecond accuracy and some randomness to the filename to avoid conflicts
                newFilename = fileNameAddition + ZonedDateTime.now().toString().substring(0, 23).replaceAll(":|\\.", "-") + "_" + UUID.randomUUID().toString().substring(0, 8) + "."
```
I think a comment could help here to understand what the `substring(0, 23)` does.